### PR TITLE
[Azure] Migrate Azure logs package to adopt ecs@mappings

### PR DESCRIPTION
**Pull request by harnish-elastic**
- Enhancement

## Proposed commit message

Command

```
  go run github.com/andrewkroh/go-examples/ecs-update@014b35dfe4c9832b51e7c909a39a48257d6a005d \
    -ecs-version=8.11.0 \
    -ecs-git-ref=v8.11.0 \
    -fields-yml-drop-ecs \
    -kibana-version=^8.13.0 \
    -drop-import-mappings \
    packages/azure
```

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

### CODE REVIEW

**Comment by ishleenk17** [comment]
As part of migrating azure package to ecs@mappings, the below 2 datastreams have pipeline failures. 

**Firewall Logs:** 
```
Issue:
azure/firewall_logs test-dnsproxyrules-raw.log:
[0] parsing field value failed: field "dns.header_flags"'s value "QR" is not one of the expected values (AA, TC, RD, RA, AD, CD, DO)
The allowed values are [these](https://www.elastic.co/guide/en/ecs/current/ecs-dns.html#field-dns-header-flags). 

Solution: Considering that qr can't be a header flag, change the pipeline logs with appropriate flags
```

**Platform Logs:** 
```
Issue: 
azure/platformlogs test-platformlogs-edgecases.log:
[0] parsing field value failed: field "event.outcome"'s value "succeeded" is not one of the allowed values (failure, success, unknown)
azure/platformlogs test-platformlogs-raw.log:
[0] parsing field value failed: field "event.outcome"'s value "succeeded" is not one of the allowed values (failure, success, unknown)

Azure platformlogs, the value for event.outcome is being set from the following processors.  [Refer](https://github.com/elastic/integrations/blob/main/packages/azure/data_stream/platformlogs/elasticsearch/ingest_pipeline/default.yml#L174-L188)

  - convert:
      field: azure.platformlogs.Status
      target_field: event.outcome
      type: string
      if: "ctx?.event?.outcome == null && ctx?.azure?.platformlogs?.Status != null && ctx?.azure?.platformlogs?.Status instanceof String && ['success', 'failure', 'unknown', 'Succeeded', 'Failed'].contains(ctx.azure?.platformlogs?.Status)"

Solution: Remove the Succeeded, Failed. But, are those codes coming from the azure platform logs directly. Removing these might break things if user is already using these values. 
```

cc: @zmoog, @muthu-mps

**Comment by muthu-mps** [comment]
> As part of migrating azure package to ecs@mappings, the below 2 datastreams have pipeline failures.
> 
> **Firewall Logs:**
> 
> ```
> Issue:
> azure/firewall_logs test-dnsproxyrules-raw.log:
> [0] parsing field value failed: field "dns.header_flags"'s value "QR" is not one of the expected values (AA, TC, RD, RA, AD, CD, DO)
> The allowed values are [these](https://www.elastic.co/guide/en/ecs/current/ecs-dns.html#field-dns-header-flags). 
> 
> Solution: Considering that qr can't be a header flag, change the pipeline logs with appropriate flags
> ```
> 

- The QR is one of the header values for the Azure Firewall log. We cannot ignore this value for the Firewall logs. 
- Not much information available, But you can find the details on the QR header value [here](https://techcommunity.microsoft.com/t5/azure-network-security-blog/enabling-central-visibility-for-dns-using-azure-firewall-custom/ba-p/2156331#:~:text=This%20includes%20the-,qr,-bit%20as%20well).

**Comment by muthu-mps** [comment]
> **Platform Logs:**
> 
> ```
> Issue: 
> azure/platformlogs test-platformlogs-edgecases.log:
> [0] parsing field value failed: field "event.outcome"'s value "succeeded" is not one of the allowed values (failure, success, unknown)
> azure/platformlogs test-platformlogs-raw.log:
> [0] parsing field value failed: field "event.outcome"'s value "succeeded" is not one of the allowed values (failure, success, unknown)
> 
> Azure platformlogs, the value for event.outcome is being set from the following processors.  [Refer](https://github.com/elastic/integrations/blob/main/packages/azure/data_stream/platformlogs/elasticsearch/ingest_pipeline/default.yml#L174-L188)
> 
>   - convert:
>       field: azure.platformlogs.Status
>       target_field: event.outcome
>       type: string
>       if: "ctx?.event?.outcome == null && ctx?.azure?.platformlogs?.Status != null && ctx?.azure?.platformlogs?.Status instanceof String && ['success', 'failure', 'unknown', 'Succeeded', 'Failed'].contains(ctx.azure?.platformlogs?.Status)"
> 
> Solution: Remove the Succeeded, Failed. But, are those codes coming from the azure platform logs directly. Removing these might break things if user is already using these values. 
> ```
> 
> cc: @zmoog, @muthu-mps

- In this case, When looking into the different Azure log integrations the `event.outcome` is not consistent from the Azure logs and when we take the sign-in logs as example the `status_code:1` is considered as `event.outcome: success` and the `status_code:0` as `event.outcome:failure`.
- Ideally, we should convert the `succeeded` and `failed` values to `success` and `failure` respectively in the ingest pipeline. 
#### Impact
- By changing the status values if the user is using them in the filter they won't see any results for succeeded or failed. Instead they will be added to the success. This is fine IMO.
- If the chart groups values by `event.outcome` this will not include the succeeded value anymore.

@zmoog - What do you think?

**Comment by ishleenk17** [comment]
> * The QR is one of the header values for the Azure Firewall log. We cannot ignore this value for the Firewall logs.

@harnish-elastic : To add this value to the expected values of DNS header flags, let's override the expected values of this field in ecs.yml, something like [this](https://github.com/elastic/integrations/pull/10223#issuecomment-2190750962) example. Let's see if things work. 




> * Ideally, we should convert the `succeeded` and `failed` values to `success` and `failure` respectively in the ingest pipeline.

@zmoog @muthu-mps : The question here is even if you convert succeeded -> success internally, will the user be expecting this status to be succeeded only?
IMO, the time it communicates the same meaning, it should not be a big deal. We can change these to expected values in the ingest pipeline. succeeded->success and failed->failure

Another option is overriding the event.outcome in ecs.yml like [this](https://github.com/elastic/integrations/pull/10223#issuecomment-2190750962) example

**Comment by harnish-elastic** [comment]
> > * The QR is one of the header values for the Azure Firewall log. We cannot ignore this value for the Firewall logs.
> 
> @harnish-elastic : To add this value to the expected values of DNS header flags, let's override the expected values of this field in ecs.yml, something like [this](https://github.com/elastic/integrations/pull/10223#issuecomment-2190750962) example. Let's see if things work.


@ishleenk17 I have checked by adding expected values like below and it worked!
```
- name: dns.header_flags
  external: ecs
  expected_values:
    - AA
    - TC
    - RD
    - RA
    - AD
    - CD
    - DO
    - QR
```

@zmoog @muthu-mps The question here is, is there a chance that we can also get the values out of the above mentioned list from azure firewall_logs?

**Comment by ishleenk17** [comment]
@tommyers-elastic : As part of the [PR](https://github.com/elastic/integrations/pull/8050), you moved the Geo ECS fields from ecs.yml to fields.yml. 
As part migrating to ecs@mappings, ideally we should remove these fields as well and use the default ecs definition of geo fields. DO you see an issue here ?

**Comment by ishleenk17** [comment]
`@zmoog @muthu-mps The question here is, is there a chance that we can also get the values out of the above mentioned list from azure firewall_logs?`

@muthu-mps @zmoog : Are there any concerns here else we will move on with these option overriding the default values for now. 

```
@tommyers-elastic : As part of the https://github.com/elastic/integrations/pull/8050, you moved the Geo ECS fields from ecs.yml to fields.yml.
As part migrating to ecs@mappings, ideally we should remove these fields as well and use the default ecs definition of geo fields. DO you see an issue here ?
```

@tommyers-elastic : WDYT ?

**Comment by muthu-mps** [comment]
> > > * The QR is one of the header values for the Azure Firewall log. We cannot ignore this value for the Firewall logs.
> > 
> > 
> > @harnish-elastic : To add this value to the expected values of DNS header flags, let's override the expected values of this field in ecs.yml, something like [this](https://github.com/elastic/integrations/pull/10223#issuecomment-2190750962) example. Let's see if things work.
> 
> @ishleenk17 I have checked by adding expected values like below and it worked!
> 
> ```
> - name: dns.header_flags
>   external: ecs
>   expected_values:
>     - AA
>     - TC
>     - RD
>     - RA
>     - AD
>     - CD
>     - DO
>     - QR
> ```
> 
> @zmoog @muthu-mps The question here is, is there a chance that we can also get the values out of the above mentioned list from azure firewall_logs?

So far we have not seen any other values other than QR.

**Comment by muthu-mps** [comment]
> `@zmoog @muthu-mps The question here is, is there a chance that we can also get the values out of the above mentioned list from azure firewall_logs?`
> 
> @muthu-mps @zmoog : Are there any concerns here else we will move on with these option overriding the default values for now.

These are the common ones and we can go-ahead with this values.

**Comment by ishleenk17** [comment]
> > `@zmoog @muthu-mps The question here is, is there a chance that we can also get the values out of the above mentioned list from azure firewall_logs?`
> > @muthu-mps @zmoog : Are there any concerns here else we will move on with these option overriding the default values for now.
> 
> These are the common ones and we can go-ahead with this values.

@harnish-elastic : Lets proceed with these then

**Comment by muthu-mps** [comment]
> > * The QR is one of the header values for the Azure Firewall log. We cannot ignore this value for the Firewall logs.
> 
> @harnish-elastic : To add this value to the expected values of DNS header flags, let's override the expected values of this field in ecs.yml, something like [this](https://github.com/elastic/integrations/pull/10223#issuecomment-2190750962) example. Let's see if things work.
> 
> > * Ideally, we should convert the `succeeded` and `failed` values to `success` and `failure` respectively in the ingest pipeline.
> 
> @zmoog @muthu-mps : The question here is even if you convert succeeded -> success internally, will the user be expecting this status to be succeeded only? IMO, the time it communicates the same meaning, it should not be a big deal. We can change these to expected values in the ingest pipeline. succeeded->success and failed->failure
> 
> Another option is overriding the event.outcome in ecs.yml like [this](https://github.com/elastic/integrations/pull/10223#issuecomment-2190750962) example

I don't think user would expect the `succeeded` value. I would also like the hear from @zmoog based on the past experiences.

**Comment by zmoog** [comment]
> > > > `@zmoog @muthu-mps The question here is, is there a chance that we can also get the values out of the above mentioned list from azure firewall_logs?`
> > > > @muthu-mps @zmoog : Are there any concerns here else we will move on with these option overriding the default values for now.
> > > 
> > > 
> > > These are the common ones and we can go-ahead with this values.
> > 
> > 
> > @harnish-elastic : Lets proceed with these then
> 
> Changes are already pushed!

I'm late to the party, but it's +1 from me.

> > The question here is even if you convert succeeded -> success internally, will the user be expecting this status to be succeeded only? IMO, the time it communicates the same meaning, it should not be a big deal. We can change these to expected values in the ingest pipeline. succeeded->success and failed->failure
> > Another option is overriding the event.outcome in ecs.yml like [this](https://github.com/elastic/integrations/pull/10223#issuecomment-2190750962) example
> 
> I don't think user would expect the `succeeded` value. I would also like the hear from @zmoog based on the past experiences.

We usually try to maintain as much backward compatibility as possible. However, the current `event.outcome` field values are not the expected ones per ECS.

Platform logs are a semi-generic data stream. The Azure Native ISV [uses](https://github.com/elastic/azuremarketplacedev/issues/152) platform logs as the default data stream when it doesn't have a specific routing. 

Aligning the `event.outcome` field to the ECS expected values (succeeded->success and failed->failure) feels like a breaking change. Since platform logs are a generic data stream, users may use this field to check for event outcomes on various log categories, in ways that go beyond what the integration offers.

To take this as the opportunity to align the expected values (succeeded->success and failed->failure) for the `event.outcome` field, we probably need to highlight this release contains breaking changes, like bumping package version from v1 to v2. I would take this opportunity. 

@jsoriano, what is the best practice to communicate breaking changes in integration releases?

**Comment by jsoriano** [comment]
> @jsoriano, what is the best practice to communicate breaking changes in integration releases?

As you mention we try to avoid them, but if needed, you could bump the major version, and include a "breaking-change" entry in the changelog.
I think this would be more than enough for this case where we are slightly changing some values for coherence with ECS and other integrations.

**Comment by zmoog** [comment]
> > @jsoriano, what is the best practice to communicate breaking changes in integration releases?

> As you mention we try to avoid them, but if needed, you could bump the major version, and include a "breaking-change" entry in the changelog. I think this would be more than enough for this case where we are slightly changing some values for coherence with ECS and other integrations.

Yeah, I try to avoid breaking changes as much as possible. In this case, aligning the `event.outcome` field value to ECS and other integrations is probably worth the cost. 

To properly mark this change as breaking, do we need to set `type` as in the following example?

```yaml
- version: "0.4.0"
  changes:
    - description: added anonymous auth config, replaced some RUM config
      type: breaking-change
      link: https://github.com/elastic/apm-server/pull/5623
```

**Comment by jsoriano** [comment]
> In this case, aligning the `event.outcome` field value to ECS and other integrations is probably worth the cost.

Agree. And if any user is affected by this change the adaptation would be easy, by modifying queries or dashboards to use the new values.

> To properly mark this change as breaking, do we need to set `type` as in the following example?

Correct.

**Comment by ishleenk17** [comment]
> To properly mark this change as breaking, do we need to set `type` as in the following example?

@harnish-elastic : Let's take care of this. 
Please resolve the CI error

**Comment by harnish-elastic** [comment]
> > To properly mark this change as breaking, do we need to set `type` as in the following example?
> 
> @harnish-elastic : Let's take care of this. Please resolve the CI error

Updated, thanks!

CI is getting failed due to SonarQube Code Analysis which can be ignored as per [this](https://github.com/elastic/integrations/pull/10225#issuecomment-2210528698) comment!

File: packages/azure/changelog.yml [hunk 1]
```diff
@@ -1,3 +1,11 @@
+- version: "1.12.0"
+  changes:
+    - description: ECS version updated to 8.11.0. Update the kibana constraint to ^8.13.0. Modified the field definitions to remove ECS fields made redundant by the ecs@mappings component template.
```
Comment (muthu-mps) [inline packages/azure/changelog.yml:3]: This changelog has multiple updates, Can we have separate descriptions for each.
- Update ECS version to 8.11.0
- Update the kibana constraint to ^8.13.0 to adopt ecs@mappings component template.

No need to add this sentence in the changelog entry,
`Modified the field definitions to remove ECS fields made redundant by the ecs@mappings component template.`

**Comment by muthu-mps** [comment]
@harnish-elastic - After making the changes, Did we test the data collection for any of the Azure logs integration?

**Review by efd6 — APPROVED** [review]
application_gateway and firewall_logs LGTM

File: packages/azure/data_stream/firewall_logs/fields/ecs.yml [hunk 2]
```diff
@@ -1,129 +1,11 @@
...
-  external: ecs
+  expected_values:
+    - AA
+    - TC
+    - RD
+    - RA
+    - AD
+    - CD
+    - DO
+    - QR
```
[the remark this answers is not in the document]
Comment (ishleenk17) [reply packages/azure/data_stream/firewall_logs/fields/ecs.yml:11]: QR (Query/Rsponse) seems to be a valid DNS header flag.
IMO, we should not remove it here, rather have it added in the allowed list. 
@jsoriano

**Review by ishleenk17 — APPROVED** [review]
Looks good!
Please address others' comments if pending.

**Comment by harnish-elastic** [comment]
> @harnish-elastic - After making the changes, Did we test the data collection for any of the Azure logs integration?

I have done the pipeline tests as Azure package is not having the system tests for platformlogs data stream. Pipeline tests are working as expected!

**Review by niraj-elastic — APPROVED** [review]
LGTM

**Review by muthu-mps — APPROVED** [review]
LGTM!

File: packages/azure/changelog.yml [hunk 3]
```diff
@@ -1,3 +1,11 @@
...
+      type: enhancement
+      link: https://github.com/elastic/integrations/pull/10224
+    - description: Updated "event.outcome" values from "Succeeded" to "success" and "Failed" to "failure".
+      type: breaking-change
+      link: https://github.com/elastic/apm-server/pull/10224
```
Comment (zmoog) [inline packages/azure/changelog.yml:8]: Fix repo name:

```suggestion
      link: https://github.com/elastic/integrations/pull/10224
```

**Review by zmoog — APPROVED** [review]
LGTM, only a minor fix to the repo name in the changelog link.

File: packages/azure/changelog.yml (hunk 1 above)
Comment (andrewkroh) [reply packages/azure/changelog.yml:3]: Separate entries sounds like a reasonable enhancement to make to `ecs-update`. The related code is at https://github.com/andrewkroh/go-examples/blob/7bd2d3beac526b97b7734b2b51433d4be54d0fb6/ecs-update/ecs-update.go#L343 if anyone wants to contribute.

**Pull request merged**